### ISSUE TITLE
Coloured Tunics cause weirdshot crashes

### ISSUE THREAD

**Issue by dotzo** [OoTRandomizer/OoT-Randomizer#751]
Depending on the colour of the tunic and the tunic item itself a weirdshot crash can occur if a "weird" frame is stumbled upon.  So far consistent known crash colour combinations are 0x5ACBCB Kokiri tunic, BQ default gold Kokiri Tunic, BQ default purple Goron Tunic.  This has not been fully explored and we do not know the underlying reason.

**Comment by dotzo** [comment]
Forgot to mention, currently this is only known to be consistent on VC.

**Comment by dotzo** [comment]
Fig and mzx did some digging and found some relevant addresses.  There are 2 sets of addresses in RAM that control tunic color and they're updated every frame.  One set is static and appears to just read from rom. The other set, below, is dynamic and changes based on if link is blinking red from damage, targetting etc.  It sits in a block of code that's inside the dlist buffer.

Addresses:
1.0
800F7AD8  Kokiri Tunic R
800F7AD9  Kokiri Tunic G
800F7ADA  Kokiri Tunic B

800F7ADB  Goron Tunic R
800F7ADC  Goron Tunic G
800F7ADD  Goron Tunic B

800F7ADE  Zora Tunic R
800F7ADF  Zora Tunic G
800F7AE0  Zora Tunic B


dlistbuffer:
11F290: Graphics Context
DLIST COMPLETE: 8017C548
01B4 WORK_DISP    : 00400 18E958 18E958 18ED58 Free: 00400
02B8 POLY_OPA_DISP: 0BF00 17CA58 17D100 187C78 Free: 0AB78
02C8 POLY_XLU_DISP: 04000 188958 188C30 18C958 Free: 03D28
02A8 OVERLAY_DISP : 02000 18C958 18C9C0 18E958 Free: 01F98

11F290: Graphics Context
DLIST COMPLETE: 8018E958
01B4 WORK_DISP    : 00400 17C548 17C550 17C948 Free: 003F8
02B8 POLY_OPA_DISP: 0BF00 16A648 16B080 175328 Free: 0A2A8
02C8 POLY_XLU_DISP: 04000 176548 176AA8 17A548 Free: 03AA0
02A8 OVERLAY_DISP : 02000 17A548 17A5E8 17C548 Free: 01F60

duh, double buffered
so 16A648 is the start of POLY_OPA_DISP

**Comment by mzxrules** [comment]
this is a copy of Krim's writeup, found here: https://pastebin.com/raw/K0AdCBNn

Function 80079F48 (1.0) is link's drawing function.
A0 = Global Context
A1 = Ptr in link's object
A2 = Ptr to link's animation ctx (link + 0x1F0)

This function will load the halfword (signed) at A2 + 0x84.
During the weirdshot, the frame that loads links tunic color, this value is 0x94AD.
The value is AND'd with 0xF, and 1 is subtracted from it (0x94AD & 0xF) - 1 = 0xC
This value is used as an index to a table at 800F7AA8 which contains a list of addresses to set segment address 8 to.

This table contains 0xB entries, and right after it is the RGB of link's tunic colors.
The value 0xRRGGBBXX is loaded Where G is the green component of the kokiri tunic color, B is the Blue component of the kokiri tunic color, and X is the Red component of the goron tunic color.  This value is AND'd with 0x00FFFFFF to get the physical memory address for segment 8

Therefore the gbi command DB060020 00GGBBXX.  By default this would be DB060020 00691B64

Later on when the display is processed this value is used to load a texture tile.  When the memory size is 4MB this command will be ignored in emulators as 0x00691B64 is outside the RDRAM Size, when the memory size is 8MB the memory at 0x00691B64 will be treated as a texture.

When the green component is changed to greater than 0x7F this value will be ignored in emulators as anything above 0x801B64 would be outside of the RDRAM size.

VC Seems to completely ignore checking the RDRAM size before attempting to load the data.  The address to load is added to the start of n64 rdram which is 0x80E74000 to get 0x81505B64.  This is a valid address on the wii and the texture is loaded.

When the green component is 0x98 we do the same addition to get address 0x817F5B64 which is still a valid address.
When the green component is 0x99 we get 0x81805B64 which is an INVALID address, the memory size of MEM1 is 24MB Which ends at 0x81800000, so we can see that 0x98 barely makes it, and 0x99 pushes that past the boundary, and this is what causes the crash.

TL;DR; Segment register 8 is set based on a table indexed by links animation + 0x84, the weirdshot indexes past this table to tunic colors.  This value is ignored by emulators (and possibly n64) when the address is past dram size, VC does not ignore the dram limitation and will crash when the green component becomes 0x99+

**Comment by r0bd0g** [comment]
mzxrules Today at 2:57 AM

random tunic colors breaks weirdshot confirmed: https://pastebin.com/raw/K0AdCBNn

**Comment by r0bd0g** [comment]
I came here to update this issue but it seems it's been done :p

**Comment by ETR-BTF** [comment]
#1698

**Comment by fenhl** [comment]
Closing per https://github.com/OoTRandomizer/OoT-Randomizer/pull/1965#issuecomment-1741758607. If anyone is still experiencing the crash on seeds both generated and patched on randomizer versions 7.1.96 or newer, please leave a comment and we will reopen this issue.
